[Doc][Core][State Observability] Adding Python SDK doc and docstring (#26997)
ray-project/ray: doc/source/ray-observability/state/state-api.rst

Now, let's see the summarized states of tasks. If it doesn't return the output immediately, retry the command.

.. tabbed:: CLI

    .. code-block:: bash

        ray summary tasks

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.common import StateResource

>         from ray.experimental.state.state_cli import format_summary_output

Reviewer: Same for `StateResource `

You can also access logs through ``ray logs`` API.

.. tabbed:: CLI

    .. code-block:: bash

        ray list actors

>         ray logs --actor-id <ACTOR_ID>

Reviewer: Can you add `# In this case, 31405554844820381c2f0f8501000000`?

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import get_log

>         actor_id = "31405554844820381c2f0f8501000000"

Reviewer: I think this <ACTOR_ID> vs concrete actor_id value thing is not consistent across all examples.

Can you just unify it to <ACTOR_ID> format? Like 

```
for line in get_log(address="http://localhost:8265", actor_id=<ACTOR_ID>): # In this case, ACTOR_ID is 31405554844820381c2f0f8501000000
```

>         for line in get_log(address="http://localhost:8265", actor_id=actor_id):

Reviewer: ```suggestion
        for line in get_log(address="http://localhost:6379", actor_id=actor_id):
```

Actually, should we just do <actor_id> as well? Also can you mention `address` should be equivalent to the address given to ray.init? Btw should we auto-resolve this for `get_log`?

You can also access logs through ``ray logs`` API.

.. tabbed:: CLI

    .. code-block:: bash

        ray list actors
        ray logs --actor-id <ACTOR_ID>

.. tabbed:: Python SDK

    .. code-block:: python

        import ray
        from ray.experimental.state.api import get_log

        # To connect to an existing ray instance if there is

>         ray.init("auto")

Author: `ray.init` needed so that I don't have to pass in the address since the python log api doesn't resolve api server address automatically.

>         ray.init("auto")

Author: I guess I could also make it a bit more verbose by passing in the api server url.

you can use ``list`` or ``get`` APIs to get more details for an individual abnormal resource.

E.g., Summarize all actors
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray summary actors

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import summarize_actors

        print(summarize_actors())

E.g., Summarize all tasks  
~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray summary tasks

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import summarize_tasks

        print(summarize_tasks())

E.g., Summarize all objects  
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note::

    By default, objects are summarized by callsite. However, callsite is not recorded by Ray by default.
    To get callsite info, set env variable `RAY_record_ref_creation_sites=1` when starting the ray cluster
    RAY_record_ref_creation_sites=1 ray start --head

.. tabbed:: CLI

    .. code-block:: bash

        ray summary objects 

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import summarize_objects

> 

Reviewer: ```suggestion
```

(Remove a new line to be consistent to other tabs)

    .. code-block:: python

        from ray.experimental.state.api import list_tasks 
        list_tasks(detail=True)

Get
---

E.g., Get a task info
~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray get tasks <task_id> 

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import get_task 
        get_task(id="<task_id>")

E.g., Get a node info
~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray get nodes <node_id> 

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import get_node 
        get_node(id="<node_id>")

Logs
----

State API also allows you to conveniently access ray logs. Note that you cannot access the logs from a dead node.
By default, the API prints log from a head node.

E.g., Get all retrievable log file names from a head node
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray logs 

.. tabbed:: Python SDK

    .. code-block:: python

        # You could get the node id / node ip from `ray list nodes` 
        from ray.experimental.state.api import list_logs 

>         list_logs(node_id="<head_node_id>")

Reviewer: ```suggestion
       # `ray logs` by default print logs from a head node. So in order to list the same logs, you should provide the head node id. You could get the node id / node ip from `ray list nodes` 
        list_logs(node_id="<head_node_id>")
```

E.g., Get a particular log file from a node
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        # You could get the node id / node ip from `ray list nodes` 
        ray logs gcs_server.out --node-id <XYZ> 

.. tabbed:: Python SDK

    .. code-block:: python

        # You could get the node id / node ip from `ray list nodes` 
        from ray.experimental.state.api import get_log 

        # Node IP could be retrieved from list_nodes() or ray.nodes()
        node_ip = "172.31.47.143" 

>         for line in get_log(address="http://localhost:8265", filename="gcs_server.out", node_ip=node_ip):

Reviewer: ```suggestion
        for line in get_log(address="<address given to ray.init>", filename="gcs_server.out", node_ip=node_ip):
```

Author: Will do auto-resolution.

            print(line)

E.g., Stream a log file from a node
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray logs -f raylet.out --node-ip 172.31.47.143

.. tabbed:: Python SDK

    .. code-block:: python

        # You could get the node id / node ip from `ray list nodes` 
        from ray.experimental.state.api import get_log 

        # Node IP could be retrieved from list_nodes() or ray.nodes()
        node_ip = "172.31.47.143" 

>         for line in get_log(address="http://localhost:8265", filename="raylet.out", node_ip=node_ip, follow=True):

Reviewer: ```suggestion
        for line in get_log(address="<address given to ray.init>, filename="raylet.out", node_ip=node_ip, follow=True):
```

            print(line)

E.g., Stream log from a pid 
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray logs --pid=<XXX> --follow

.. tabbed:: Python SDK

    .. code-block:: python

        # You could get the node id / node ip from `ray list nodes` 
        import ray
        from ray.experimental.state.api import get_log 

        # Node IP could be retrieved from list_nodes() or ray.nodes()
        node_ip = "172.31.47.143" 
        # You could get the pid of the worker running the actor easily when output
        # of worker being directed to the driver (default)
        pid = "318158" 
        # The loop will block with `follow=True`

>         for line in get_log(address="http://localhost:8265", pid=pid, node_ip=node_ip, follow=True):

Reviewer: ```suggestion
        for line in get_log(address="<address given to ray.init>", pid=pid, node_ip=node_ip, follow=True):
```

        for line in get_log(filename="gcs_server.out", node_id=<NODE_ID>):
            print(line)

E.g., Stream a log file from a node
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        # You could get the node id / node ip from `ray list nodes` 
        ray logs -f raylet.out --node-ip <NODE_IP> 

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import get_log 

        # Node IP could be retrieved from list_nodes() or ray.nodes()
        for line in get_log(filename="raylet.out", node_ip=<NODE_IP>, follow=True):
            print(line)

E.g., Stream log from a pid 
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray logs --pid=<PID> --follow

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import get_log 

        # Node IP could be retrieved from list_nodes() or ray.nodes()
        # You could get the pid of the worker running the actor easily when output
        # of worker being directed to the driver (default)
        # The loop will block with `follow=True`

>         for line in get_log(pid=<PID>, node_ip=<NODE_IP>, follow=True):

Author: This will no longer be relevant after the logging refactor (which allows auto-population of `node_ip` and `node_id`). See [this](https://github.com/ray-project/ray/pull/27069/files#diff-1f9ff1a83a3da4525dd3be04b6aa2fb29e862e6484f46e6676ce38652373291bR1131)

        from ray.experimental.state.api import summarize_objects

        # To get callsite info, set env variable `RAY_record_ref_creation_sites=1` when starting the ray cluster
        # RAY_record_ref_creation_sites=1 ray start --head
        print(format_summary_output(summarize_objects(), resource=StateResource.OBJECTS))

List
----

Get a list of resources, possible resources include: 

- :ref:`Actors <actor-guide>`, e.g., actor id, state, pid, death_cause. (:ref:`output schema <state-api-schema-actor>`)
- :ref:`Tasks <ray-remote-functions>`, e.g., name, scheduling state, type, runtime env info (:ref:`output schema <state-api-schema-task>`)
- :ref:`Objects <objects-in-ray>`, e.g., object id, callsites, reference types. (:ref:`output schema <state-api-schema-obj>`)
- :ref:`Jobs <jobs-overview>`, e.g., start/end time, entrypoint, status. (:ref:`output schema <state-api-schema-job>`)
- :ref:`Placement Groups <ray-placement-group-doc-ref>`, e.g., name, bundles, stats. (:ref:`output schema <state-api-schema-pg>`)
- Nodes (Ray worker nodes), e.g., node id, node ip, node state. (:ref:`output schema <state-api-schema-node>`)

> - Workers (Ray worker processes), e.g., worker id, type, exit type and details. (:ref:`output schema <state-api-schema-worker>`)

Reviewer: nice :)!

- :ref:`Runtime environments <runtime-environments>`, e.g., runtime envs, creation time, nodes (:ref:`output schema <state-api-schema-runtime-env>`)

E.g., List all nodes 
~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray list nodes 

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import list_actors 
        list_actors()

E.g., List all placement groups 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray list placement-groups

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import list_placement_groups 
        list_placement_groups()

 
E.g., List local referenced objects created by a process
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tip:: You can list resources with one or multiple filters: using `--filter` or `-f`

.. tabbed:: CLI

    .. code-block:: bash

        ray list objects -f pid=12345 -f reference_type=LOCAL_REFERENCE

.. tabbed:: Python SDK

    .. code-block:: python

        from ray.experimental.state.api import list_objects 

>         list_objects(filters=[("pid", "=", "12345"), ("reference_type", "=", "LOCAL_REFERENCE")])

Author: Yeah, either works. I think this is either way, args are sent through http requests, which are treated as strings at certain point.

By default, the API prints log from a head node.

> E.g., Get all retrievable log file names

Reviewer: Can you say from a head node?

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray logs 

.. tabbed:: Python SDK

    .. code-block:: python

        # You could get the node id / node ip from `ray list nodes` 
        from ray.experimental.state.api import list_logs 

>         list_logs(node_id="<node_id>")

Reviewer: maybe call it head_node_id?

Reviewer: (since that's the default)

E.g., Get a particular log file from a node
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        # You could get the node id / node ip from `ray list nodes` 
        ray logs gcs_server.out --node-id <XYZ> 

.. tabbed:: Python SDK

    .. code-block:: python

        # You could get the node id / node ip from `ray list nodes` 
        import ray
        from ray.experimental.state.api import get_log 

        # To connect to an existing ray instance if there is
        ray.init("auto")
        # Node IP could be retrieved from list_nodes() or ray.nodes()
        node_ip = "172.31.47.143" 

>         for l in get_log(filename="gcs_server.out", node_ip=node_ip):

Reviewer: ```suggestion
        for line in get_log(filename="gcs_server.out", node_ip=node_ip):
```

            print(l)

E.g., Stream a log file from a node
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabbed:: CLI

    .. code-block:: bash

        ray logs -f raylet.out --node-ip 172.31.47.143

.. tabbed:: Python SDK

    .. code-block:: python

        # You could get the node id / node ip from `ray list nodes` 
        import ray
        from ray.experimental.state.api import get_log 

        # To connect to an existing ray instance if there is
        ray.init("auto")
        # Node IP could be retrieved from list_nodes() or ray.nodes()
        node_ip = "172.31.47.143" 

>         for l in get_log(filename="raylet.out", node_ip=node_ip, follow=True):

Reviewer: Can you mention this is blocking forever?

>         for l in get_log(filename="raylet.out", node_ip=node_ip, follow=True):

Reviewer: ```suggestion
        for line in get_log(filename="raylet.out", node_ip=node_ip, follow=True):
```

>             print(l)

Reviewer: ```suggestion
            print(line)
```
